File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function kthSmallest(root: TreeNode | null, k: number): number {
  let count=0;
  let result = Infinity;
  function dfs(node) {
    if (!node) return;
    dfs(node.left);
    count++
    if (count === k) {
      result = node.val;
      return;
    }
    dfs(node.right);
  } 
  dfs(root)
  return result;

  // if (!root) return -1;
  // const stack = [];
  // let curr = root;
  // let count = 0;   
  // while ( curr || stack.length != 0 ) {
  //   while (curr) {
  //     stack.push(curr)
  //     curr = curr?.left;
  //   }
  //   const node = stack.pop();
  //   count++
  //   if (count === k) return node.val;
  //   curr = curr?.right;
  // } 
}

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        min_heap = []
        def dfs(node: Optional[TreeNode]):
          if not node:
            return
          heappush(min_heap, -node.val)
          if len(min_heap) > k:
            heappop(min_heap)
          if not node.right == None:
            dfs(node.right)
          if not node.left == None:
            dfs(node.left)
          return
        
        dfs(root)
        return -min_heap[0]